src/linker/Linker.Dataflow/ReflectionMethodBodyScanner.cs

```diff
@@ -164,24 +166,38 @@ protected override void WarnAboutInvalidILInMethod (MethodBody method, int ilOff
 			Debug.Fail ("Invalid IL or a bug in the scanner");
 		}
 
-		MethodReturnValue CreateMethodReturnValue (MethodReference method, DynamicallyAccessedMemberTypes dynamicallyAccessedMemberTypes = DynamicallyAccessedMemberTypes.None)
-		{
-			return new MethodReturnValue (ResolveToTypeDefinition (method.ReturnType), dynamicallyAccessedMemberTypes, method.MethodReturnType);
-		}
+		MethodReturnValue CreateMethodReturnValue (MethodDefinition method, DynamicallyAccessedMemberTypes dynamicallyAccessedMemberTypes)
+			=> new (ResolveToTypeDefinition (method.ReturnType), method, dynamicallyAccessedMemberTypes);
+
+		MethodReturnValue CreateMethodReturnValue (MethodDefinition method)
+			=> new (
+				ResolveToTypeDefinition (method.ReturnType),
+				method,
+				_context.Annotations.FlowAnnotations.GetReturnParameterAnnotation (method));
+
+		LeafValueWithDynamicallyAccessedMemberNode GetMethodParameterValue (MethodDefinition method, int parameterIndex, DynamicallyAccessedMemberTypes dynamicallyAccessedMemberTypes)
+			=> GetMethodParameterValueInternal (method, parameterIndex, dynamicallyAccessedMemberTypes);
 
-		protected override ValueNode GetMethodParameterValue (MethodDefinition method, int parameterIndex)
+		protected override LeafValueWithDynamicallyAccessedMemberNode GetMethodParameterValue (MethodDefinition method, int parameterIndex)
+			=> GetMethodParameterValueInternal (method, parameterIndex, _context.Annotations.FlowAnnotations.GetParameterAnnotation (method, parameterIndex));
+
+		LeafValueWithDynamicallyAccessedMemberNode GetMethodParameterValueInternal (MethodDefinition method, int parameterIndex, DynamicallyAccessedMemberTypes dynamicallyAccessedMemberTypes)
 		{
-			DynamicallyAccessedMemberTypes memberTypes = _context.Annotations.FlowAnnotations.GetParameterAnnotation (method, parameterIndex);
+			if (method.HasImplicitThis ()) {
+				if (parameterIndex == 0)
+					return new MethodThisParameterValue (method, dynamicallyAccessedMemberTypes);
 
-			var staticType =
-				!method.HasImplicitThis () ? ResolveToTypeDefinition (method.Parameters[parameterIndex].ParameterType) :
-				parameterIndex == 0 ? method.DeclaringType :
-				ResolveToTypeDefinition (method.Parameters[parameterIndex - 1].ParameterType);
+				parameterIndex--;
+			}
 
-			return new MethodParameterValue (staticType, parameterIndex, memberTypes, DiagnosticUtilities.GetMethodParameterFromIndex (method, parameterIndex));
+			return new MethodParameterValue (
+				ResolveToTypeDefinition (method.Parameters[parameterIndex].ParameterType),
```

Comment:
I actually tried before, but it needs context - or rather it would need to duplicate the functionality of `ResolveToTypeDefinition` which includes calling into `LinkContext`. I thought that this way it's a bit less weird.